### DISCUSSION TITLE
Direct SQL upload very slow?

### DISCUSSION THREAD

**Discussion by laurentS** [supabase/supabase#577] [Questions] [answered]
Hi folks!
I'm trying to upload data into a table and I'm surprised at how slow it is. I'd love some tips on what is making it that slow:
- I load a CSV file into python/pandas, clean it up and then use pandas.to_sql() to connect to supabase (it uses sqlalchemy under the hood, so all of that should be as stable as it gets)
- my test dataset has 100 rows and 3 columns (1 int and 2 varchar that don't exceed about 50 characters each)
- The logs show that sqlalchemy creates the table:
```
CREATE TABLE mytable (
	id BIGINT, 
	"name" TEXT, 
	"reference" TEXT
)
```
- and then uploads the data:
 ```COMMIT
BEGIN (implicit)
INSERT INTO mytable (id, "name", "reference") VALUES (%(id)s, %(name)s, %(reference)s)  ... <the data>
COMMIT
```
- I've varied the number of rows to upload and the `INSERT` takes around 200ms per row, no matter what (so 20 seconds for 100 rows). This does not feel right, as there are no constraints on the table. (when testing on a local postgres on my laptop, the INSERT takes 0.35ms per row, that's 570 times faster). The string representing the data is less than 10kB, so upload time cannot explain this) 
- As my data is around 200k rows with 20+ columns, I'm anticipating north of 12h of upload :sob: and then the other tables...

Are there some settings I can change to speed this up, or is there a different approach I should be using?

**Comment by kiwicopple** [comment]
It might be related to replication. Can you please try this?

```sql
begin; 
  -- remove the realtime publication
  drop publication if exists supabase_realtime; 

  -- re-create the publication but don't enable it for any tables
  create publication supabase_realtime;  
commit;
```

then let us know if uploading is faster?

**Reply by laurentS** [reply]
I tried the following (confirming here as I'm not 100% sure about how to use your query):
- ran your query through the supabase query editor. I just get "success" but no return value
- ran my upload code again. I see no different in the time the INSERT takes (the actual time was closer to 190ms than the 200ms I mentioned before, and it's still very much the same). I tried dropping/recreating the table after your query as well, but it doesn't seem to make any difference.

I also tried to upload a single column of INT, or a single column of VARCHAR, but the resulting time stays the same in all cases.

**Comment by steve-chavez** [comment, marked as answer]
It's possible for an individual insert to take that long, specially if you created your db in a region far away from yours. I'm seeing up to `400ms` for a db in asia when I'm in south america:

```sql
> psql postgresql://youruri

> \timing on

> insert into projects values (1, 'project-1');
Time: 426.272 ms
```

@laurentS What I'd suggest here is to batch your inserts and not send each one individually. Seems this is possible in sqlalchemy, see [bulk operations](https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#bulk-operations).

Another option would be to construct a single INSERT with many values manually, like:

```sql
insert into projects values 
(1, 'project-1'),
(2, 'project-2'),
(3, 'project-3');
-- ...
-- ...
```

**Reply by laurentS** [reply]
Hi @steve-chavez thanks for these pointers. I was convinced that pandas was constructing a single INSERT because of how the logs looked, but it was, in fact, not doing it. Putting this down here, in case someone else hits the same issue with `pandas.to_sql(...)`, using the `method="multi"` argument does what you suggested and inserts multiple values in one INSERT. With it, the insert time drops from about 200ms per row down to 0.5ms (and a bit more, close to ping time if inserting just a single row).
I realise this is not a supabase issue at all, apologies for the noise! But thanks for the help, much appreciated!
